### DISCUSSION TITLE
preparing multiple fabric images and document printing them - shows empty window

### DISCUSSION THREAD

**Discussion by emre-ozgun** [fabricjs/fabric.js#8384] [Q&A]
I can successfully construct images from an array of canvases. When I print, It shows a blank screen (HTML Object Element)
What part of my logic is flawed ? I'd be glad to share additional information on request :)

Here is the stack issue: https://stackoverflow.com/questions/74099956/fabricjs-canvas-converting-multiple-fabricjs-canvases-to-images-via-loop-a

```
  const handleCustomPrint = () => {

    const fabricImagesToBePrinted = document.createElement('div');

    for (let i = 0; i < canvasArrayToBePrinted.length; i++) {
      const canvas = new fabric.Canvas(`${i}`)
      canvas.loadFromJSON(canvasArrayToBePrinted[i], () => {
        const img = canvas.toDataURL();
        const imgElement = document.createElement('img');
        imgElement.src = img;
        fabricImagesToBePrinted.appendChild(imgElement);
      });
    }


    var mywindow = window.open('', 'PRINT', `height=${canvasArrayToBePrinted.length * pageSize.heightPX}, width=${pageSize.widthPX}`);

    mywindow.document.write('<html><head><title>' + document.title + '</title>');
    mywindow.document.write('</head><body >');
    mywindow.document.write('<h1>' + document.title + '</h1>');
    mywindow.document.write(fabricImagesToBePrinted.children);
    mywindow.document.write('</body></html>');

    mywindow.document.close(); // necessary for IE >= 10
    mywindow.focus(); // necessary for IE >= 10*/

    mywindow.print();
    mywindow.close();

  };
```

**Comment by iocoker** [comment]
Everything happens synchronously so fast, I doubt the image is loaded before exiting this line `canvas.loadFromJSON(canvasArrayToBePrinted[i], () => {` to this ` });` and out of the loop.

Don't you want to use a promise instead?

**Reply by emre-ozgun** [reply]
Hey I resolved the issue with a different solution. This image conversion was actually not that reliable as you pointed out. I just create canvases in a React component (potentially N images, N > 3000) And I keep a 'loadedImagesCount' state in parent to ensure that all the images load (loadedImagesCount === totalImagesCount),  AND THEN initate the print. Synchronization problem solved !
